tests/src/Kernel/CollaboraMediaAccessTest.php

```diff
@@ -0,0 +1,322 @@
+    }
+
+    /**
+     * Asserts Collabora media access for a user account.
+     *
+     * @param array<string, list<'preview', 'edit'> $expected
+     *   Expected access matrix.
+     *   The keys identify media entities that are created in this test.
+     *   The values identify operations.
+     * @param \Drupal\Core\Session\AccountInterface $account
+     *   Account to check access for.
+     * @param string $message
+     *   Message for the assertion.
+     */
+    protected function assertCollaboraMediaAccess(array $expected, AccountInterface $account, string $message): void {
+        $other_user = $this->createUser();
+        $entities = [
+            "published document" => $this->createMediaEntity('document', [
+                'uid' => $other_user->id(),
+            ]),
+            "unpublished document" => $this->createMediaEntity('document', [
+                'uid' => $other_user->id(),
+                'status' => 0,
+            ]),
+            "own published document" => $this->createMediaEntity('document', [
+                'uid' => $account->id(),
+            ]),
+            "own unpublished document" => $this->createMediaEntity('document', [
+                'uid' => $account->id(),
+                'status' => 0,
+            ]),
+        ];
+
+        // Test $entity->access() with different operations on all entities.
+        $operations = [
+            'preview' => 'preview in collabora',
+            'edit' => 'edit in collabora',
+        ];
+        $actual_entity_access = [];
+        foreach ($entities as $entity_key => $entity) {
+            foreach ($operations as $operation_key => $operation) {
+                $has_entity_access = $entity->access($operation, $account);
+                if ($has_entity_access) {
+                    $actual_entity_access[$entity_key][] = $operation_key;
+                }
+            }
+        }
+        $this->assertSameYaml(
+            $expected,
+            $actual_entity_access,
+            'Entity access: ' . $message,
+        );
+
+        // Test path access.
+        // The result is expected to be exactly the same, due to how the route
+        // access is configured.
+        // Testing the paths like this introduces some level of redundancy or
+        // duplication, but it is cheap and easy, so for now this is what we do.
+        $sprintf_path_patterns = [
+            'preview' => '/cool/view/%s',
+            'edit' => '/cool/edit/%s',
+        ];
+        $actual_path_access = [];
+        foreach ($entities as $entity_key => $entity) {
+            foreach ($sprintf_path_patterns as $pattern_key => $sprintf_path_pattern) {
+                $path = sprintf($sprintf_path_pattern, $entity->id());
+                $has_path_access = Url::fromUserInput($path)->access($account);
+                if ($has_path_access) {
+                    $actual_path_access[$entity_key][] = $pattern_key;
+                }
+            }
+        }
```

Comment:
Actually, I think you don't need to duplicate loops, since you have to loop same entities and operations.
Just need to store results in the different variables and check after.

This would reduce the duplicated code and keep doing path checks.

And also have descriptive messages for each.